File: mergedialog.cpp
#include "mergedialog.h"
#include <QDebug>

const int MergeDialog::M_MAX_FILENUM=50;

MergeDialog::MergeDialog(QWidget *parent):
    QDialog(parent),
    m_model(1,1),
    m_selected(false)
{
    m_ui.setupUi(this);
    this->setWindowTitle("Merging Files Dialog");
    m_ui.m_fileListView->setModel(&m_model);
    connect(m_ui.m_selectButton,&QPushButton::clicked,this,&MergeDialog::onSelectSourceFiles);
    connect(m_ui.m_saveButton,&QPushButton::clicked,this,&MergeDialog::onMerging);
    connect(m_ui.m_quitButton,&QPushButton::clicked,this,&MergeDialog::close);
}

void MergeDialog::onSelectSourceFiles()
{
    QStringList file_list=QFileDialog::getOpenFileNames(this,tr("Open data file"), "D:\\",
                                  tr("Data Files (*.txt *.csv *.dat);;All Files (*.*)"));
    int rows=file_list.size();
    if(rows>0){
        m_model.setRowCount(rows);
        for(int i=0;i<rows;i++){
            m_model.setData(m_model.index(i,0,QModelIndex()),file_list.at(i));
        }
        m_selected=true;
    }
}

void MergeDialog::onMerging()
{
    if(m_selected){
        QString savefile=QFileDialog::getSaveFileName(this,tr("Open data file"), "D:\\",
                                                      tr("Data Files (*.csv);;All Files (*.*)"));
        if(savefile.trimmed()!=""){
            int button_id=m_ui.m_spanCheckGroup->checkedId();//defultly,1st ID is -2,2nd is -3...;
            if(-2==button_id){//column span
                columnMerging(savefile);
            }else if(-3==button_id){//row span
                ;
            }
        }
    }else{
        QMessageBox::warning(this,"Warning","soure files have not been selected");
    }
}

void MergeDialog::columnMerging(QString savefile)
{
    if(m_selected && savefile!=""){
        QFile fs_save(savefile);
        if (!fs_save.open(QIODevice::WriteOnly | QIODevice::Text)) {
             QMessageBox::warning(this,"warning",QString("cannot create file:%1").arg(savefile));
             return;
         }
        QTextStream out(&fs_save);
        QFile fs_selects[M_MAX_FILENUM];
        QTextStream* stream_selects[M_MAX_FILENUM];
        int filenum=0;
        for(int i=0;i<m_model.rowCount() && filenum<M_MAX_FILENUM;i++){
            QString selectfile=m_model.data(m_model.index(i,0,QModelIndex())).toString();
            fs_selects[filenum].setFileName(selectfile);
            if(!fs_selects[filenum].open(QIODevice::ReadOnly | QIODevice::Text)){
                QMessageBox::warning(this,"warning",QString("cannot open file:%1").arg(selectfile));
                continue;
            }
            stream_selects[filenum]=new QTextStream(&fs_selects[i]);
            filenum++;
        }
        if(filenum<1){
            QMessageBox::warning(this,"warning","no file be opened");
            return;
        }
        QString out_line="";
        QString in_line;
        QStringList values;
        int head_row=m_ui.m_headrowEdit->text().toInt()-1;
        int data_startrow=m_ui.m_dataStartRowEdit->text().toInt()-1;
        int x_column=m_ui.m_XColumnEdit->text().toInt()-1;
        int head_startcolumn=m_ui.m_headStartColumnEdit->text().toInt()-1;
        int ycolumn_step=m_ui.m_stepEdit->text().toInt();
        int ystart_column=m_ui.m_YStartColumnEdit->text().toInt()-1;
        int delimiter_check=m_ui.m_delimiterCheckGroup->checkedId();
        QString delimiter=",";
        if(delimiter_check==-3)
            delimiter=";";
        else if(delimiter_check==-4)
            delimiter=" ";
        else if(delimiter_check==-5)
            delimiter="\t";
        else if(delimiter_check==-6){
            if(m_ui.m_otherEdit->text().size()>=1)
                delimiter=m_ui.m_otherEdit->text();
        }
        int row_pos=0;
        int ycount_list[M_MAX_FILENUM];
        for(int i=0;i<filenum;i++){
            for(;row_pos<head_row;row_pos++){
                if(!stream_selects[i]->atEnd()){
                    stream_selects[i]->readLine();
                }
            }
            if(stream_selects[i]->atEnd())
                continue;
            in_line=stream_selects[i]->readLine();
            values = QString(in_line).split(delimiter,QString::KeepEmptyParts);
            if(x_column<values.size()){
                out_line.append(values.at(x_column)+delimiter);
            }
            ycount_list[i]=1;
            for(int j=head_startcolumn;j<values.size();j+=ycolumn_step){
                out_line.append(values.at(j)+delimiter);
                ++ycount_list[i];
            }
        }
        row_pos++;
        if(out_line.size()>=delimiter.size()){
            out_line.chop(delimiter.size());
        }
        out<<out_line<<"\n";
        for(;row_pos<data_startrow;row_pos++){
            for(int i=0;i<filenum;i++){
                if(!stream_selects[i]->atEnd())
                   in_line=stream_selects[i]->readLine();
            }
        }
        bool flag=true;
        while(flag){
            out_line="";
            for(int i=0;i<filenum;i++){
                int num=0;
                if(stream_selects[i]->atEnd()){
                    flag=false;
                }else{
                    flag=true;
                    in_line=stream_selects[i]->readLine();
                    values = QString(in_line).split(delimiter,QString::KeepEmptyParts);
                    if(x_column<values.size()){
                        out_line.append(values.at(x_column)+delimiter);
                        ++num;
                    }
                    for(int j=ystart_column;j<values.size();j+=ycolumn_step){
                        out_line.append(values.at(j)+delimiter);
                        ++num;
                    }
                }
                if(num<ycount_list[i]){
                    out_line.append(delimiter.repeated(ycount_list[i]-num));
                }
            }
            if(out_line.size()>=delimiter.size()){
                out_line.chop(delimiter.size());
            }
            out<<out_line<<"\n";
        }
        for(int i=0;i<filenum;i++){
           fs_selects[i].close();
           delete stream_selects[i];
        }
        fs_save.close();
    }
}

File: MergeDialog.py
# -*- coding: utf-8 -*-

import sys
import pandas
from PyQt5.QtCore import Qt,QCoreApplication,QModelIndex
from PyQt5.QtWidgets import QApplication,QDialog,QMessageBox,QFileDialog
from PyQt5.QtGui import QStandardItemModel
from Ui_MergeDialog import Ui_MergeDialog

class MergeDialog(QDialog):
    MAX_FILENUM=50
    MAX_READLINE=10000
    
    def __init__(self,parent=None):
        super(QDialog,self).__init__(parent)
        self.datetime_formats={
            'yyyy/mm/dd hh:mm:ss':'%Y/%m/%d %H:%M:%S',
            'yyyy/mm/dd h:m:s':'%Y/%m/%d %H:%M:%S',
            'yyyy/mm/dd hh:mm':'%Y/%m/%d %H:%M',
            'yyyy/mm/dd h:m':'%Y/%m/%d %H:%M',
            'yy/mm/dd hh:mm:ss':'%y/%m/%d %H:%M:%S',
            'yy/mm/dd h:m:s':'%y/%m/%d %H:%M:%S',
            'yy/mm/dd hh:mm':'%y/%m/%d %H:%M',
            'yy/mm/dd h:m':'%y/%m/%d %H:%M',
            'yyyy-mm-dd hh:mm:ss':'%Y-%m-%d %H:%M:%S',
            'yyyy-mm-dd h:m:s':'%Y-%m-%d %H:%M:%S',
            'yyyy-mm-dd hh:mm':'%Y-%m-%d %H:%M',
            'yyyy-mm-dd h:m':'%Y-%m-%d %H:%M',
            'yy-mm-dd hh:mm:ss':'%y-%m-%d %H:%M:%S',
            'yy-mm-dd h:m:s':'%y-%m-%d %H:%M:%S',
            'yy-mm-dd hh:mm':'%y-%m-%d %H:%M',
            'yy-mm-dd h:m':'%y-%m-%d %H:%M'
        }
        self.m_selected=False
        self.ui=Ui_MergeDialog()
        self.ui.setupUi(self)
        self.ui.retranslateUi(self)
        self.initUi()
        self.setWindowTitle("Merging Source Files")
        self.m_model=QStandardItemModel(1,1)
        self.ui.m_fileListView.setModel(self.m_model)
        self.setupConnect()
    
    def initUi(self):
        coding_list=["utf_8","utf_16","latin_1","cp65001","ascii",
            "gb2312","gb18030","gbk","big5","big5hkscs"
        ]
        self.ui.m_srcEncodingCBBox.addItems(coding_list)
        self.ui.m_srcEncodingCBBox.setCurrentIndex(0)
        self.ui.m_destEncodingCBBox.addItems(coding_list)
        self.ui.m_destEncodingCBBox.setCurrentIndex(0)
        self.ui.m_datetimeFormatCBBox.addItems([key for key in self.datetime_formats])
        self.ui.m_datetimeFormatCBBox.setCurrentIndex(0)


    def setupConnect(self):
        self.ui.m_selectButton.clicked.connect(self.onSelectSrcFiles)
        self.ui.m_saveButton.clicked.connect(self.onMerging)
        self.ui.m_quitButton.clicked.connect(self.close)
        self.ui.m_XisDateTimeCheck.stateChanged.connect(
            lambda state:self.ui.m_datetimeFormatCBBox.setEnabled(state)
        )
        self.ui.m_rowSpanCheck.stateChanged.connect(
            lambda state:(self.ui.m_headrowEdit.setEnabled(not state) 
                ,self.ui.m_dataStartRowEdit.setEnabled(not state)
                ,self.ui.m_XColumnEdit.setEnabled(not state)
                ,self.ui.m_YStartColumnEdit.setEnabled(not state)
                ,self.ui.m_stepEdit.setEnabled(not state)              
                ,self.ui.m_offsetDataHeadEdit.setEnabled(not state)
            )
        )


    def onSelectSrcFiles(self):
        (file_list,_)=QFileDialog.getOpenFileNames(self,
                "open data files", 
                "./", 
                "Data Files (*.txt *.csv *.dat);;All Files (*.*)"            
        )
        if(len(file_list)>0):
            self.m_model.setRowCount(len(file_list))
            i=0
            for filename in file_list:
                self.m_model.setData(self.m_model.index(i,0,QModelIndex()),filename)
                i=i+1
            self.m_selected=True

    def onMerging(self):
        if(self.m_selected):
            save_file,_=QFileDialog.getSaveFileName(
                self,
                "Open data file", 
                "./",
                "Data Files (*.csv);;All Files (*.*)"
            )
            if(save_file!=""):
                if(self.ui.m_colmunSpanCheck.isChecked()):#column span
                    self.columnMerging(save_file)
                elif(self.ui.m_rowSpanCheck.isChecked()):
                    self.rowMerging(save_file)
            else:
                QMessageBox.warning(self,"Warning!","please select merging file!")
        else:
            QMessageBox.warning(self,"Warning","source files have not been selected")

    def columnMerging(self,save_file):
        head_row=self.ui.m_headrowEdit.value()-1
        data_startrow=self.ui.m_dataStartRowEdit.value()-1
        x_column=self.ui.m_XColumnEdit.value()-1
        ycolumn_step=self.ui.m_stepEdit.value()
        ystart_column=self.ui.m_YStartColumnEdit.value()-1
        offset_data_head=self.ui.m_offsetDataHeadEdit.value() #数据相对字头的列偏移量，1为右偏1，-1为左偏1
        src_coding=self.ui.m_srcEncodingCBBox.currentText()
        dest_coding=self.ui.m_destEncodingCBBox.currentText()
        delimiter=","
        if(self.ui.m_semiCheck.isChecked()):
            delimiter=";"
        elif(self.ui.m_spaceCheck.isChecked()):
            delimiter=" "
        elif(self.ui.m_tabCheck.isChecked()):
            delimiter="\t"
        elif(self.ui.m_otherCheck.isChecked()):
            ss=self.ui.m_otherEdit.text()
            if(len(ss)>=1):
                delimiter=ss
        files_nume=self.m_model.rowCount()
        for i in range(files_nume):
            filename=self.m_model.data(self.m_model.index(i,0,QModelIndex()))
            mycolumns=pandas.read_csv(filename,
                sep=delimiter,
                encoding=src_coding,
                skipinitialspace=True,
                skip_blank_lines=True,
                skiprows=head_row,
                header=0,
                nrows=0
            ).columns
            use_cols=[i for i in range(ystart_column,len(mycolumns),ycolumn_step)]
            use_cols.insert(0,x_column)
            myindex_col=None
            if(self.ui.m_XasIndexCheck.isChecked()):
                myindex_col=x_column 
            data=pandas.read_csv(filename,
                sep=delimiter,
                encoding=src_coding,
                skipinitialspace=True,
                skip_blank_lines=True,
                skiprows=data_startrow,
                header=None,
                index_col=myindex_col,
                usecols=use_cols,
                nrows=self.MAX_READLINE
            )
            data.index.name="time"
            if(self.ui.m_XasIndexCheck.isChecked()):
                use_cols=use_cols[1:]
            data.columns=mycolumns[use_cols]
            if(self.ui.m_XisDateTimeCheck.isChecked()):
                mydatetime_format=self.datetime_formats[
                    self.ui.m_datetimeFormatCBBox.currentText()
                ]
                data.index=pandas.to_datetime(data.index,format=mydatetime_format)
            print(data)

    def rowMerging(self,save_file):
        QMessageBox.warning(self,"Warning!",save_file)        


if __name__=='__main__':
    QCoreApplication.setAttribute(Qt.AA_EnableHighDpiScaling) #>=Qt5.6:Enables high-DPI scaling in Qt on supported platforms (see also High DPI Displays
    app=QApplication(sys.argv)
    w=MergeDialog()
    w.show()
    sys.exit(app.exec_())


